File: killstreek/src/main/java/dev/kkorolyov/killstreek/media/HealthBar.java
package dev.kkorolyov.killstreek.media;

import dev.kkorolyov.killstreek.component.Health;
import dev.kkorolyov.pancake.platform.math.Vector;
import dev.kkorolyov.pancake.platform.media.Renderable;

import javafx.scene.canvas.GraphicsContext;
import javafx.scene.paint.Color;
import javafx.scene.paint.Paint;

/**
 * A rendered health bar.
 */
public class HealthBar implements Renderable {
	private static final Vector ORIENTATION_OFFSET = new Vector();

	private final Health health;
	private final Vector size;

	/**
	 * Constructs a new health bar.
	 * @param health health rendered by this bar
	 */
	public HealthBar(Health health, Vector size) {
		this.health = health;
		this.size = size;
	}

	@Override
	public void render(GraphicsContext g, Vector position) {
		Paint initialFill = g.getFill();
		Paint initialStroke = g.getStroke();
		double initialWidth = g.getLineWidth();

		float x = position.getX();
		float y = position.getY();
		float w = size.getX() * health.getPercentage();
		float h = size.getY();

		g.setFill(Color.RED);
		g.fillRect(x, y, w, h);

		g.setStroke(Color.DARKGRAY);
		g.setLineWidth(2);
		g.strokeRect(x, y, w, h);

		g.setFill(initialFill);
		g.setStroke(initialStroke);
		g.setLineWidth(initialWidth);
	}

	@Override
	public Vector size() {
		return size;
	}

	@Override
	public Vector getOrientationOffset() {
		return ORIENTATION_OFFSET;
	}
}

File: killstreek/src/main/kotlin/dev/kkorolyov/killstreek/media/HealthBar.kt
package dev.kkorolyov.killstreek.media

import dev.kkorolyov.killstreek.component.Health
import dev.kkorolyov.pancake.platform.math.Vector2
import dev.kkorolyov.pancake.platform.media.graphic.RenderTransform
import dev.kkorolyov.pancake.platform.media.graphic.Renderable
import dev.kkorolyov.pancake.platform.media.graphic.shape.Box
import dev.kkorolyov.pancake.platform.media.graphic.shape.Shape.Color
import dev.kkorolyov.pancake.platform.plugin.RenderMedium

/**
 * A rendered health bar.
 */
class HealthBar(
	/** rendered health */
	private val health: Health,
	/** bar size in {@code px} */
	private val size: Vector2,
	/** medium to render on */
	renderMedium: RenderMedium
) : Renderable {
	private val box: Box = renderMedium.box.apply {
		fill = Color.RED
		stroke = Color.GRAY
		strokeWidth = 2.0
	}

	override fun render(transform: RenderTransform) {
		box.size.let {
			it.x = size.x * health.percentage
			it.y = size.y
		}
		box.render(transform)
	}
}
